Full documentation of AZSL & AZSLc. (#1158)
o3de/o3de.org: content/docs/atom-guide/dev-guide/shaders/azsl/unbounded-arrays.md

---
linktitle: Unbounded Arrays in AZSL
title: AZSL, Binding Rules For Unbounded Arrays
description: There are limitations into how many and what kind of Unbounded Arrays can be declared. 
weight: 100
---

> AZSL supports declaration of Unbounded Arrays inside `ShaderResourceGroup`s, but there are limitations into how many and what kind of Unbounded Arrays can be declared. The shader compiler, **AZSLc**, makes sure those rules are not violated.  

Author: Fixed.

> Here is a summary of the rules and limitations per platform (DX12, Vulkan and Metal). '--use-spaces' and '--unique-idx' are command line arguments for **AZSLc** that are used for each platform.  

Author: Fixed.

> | --use-spaces   | --unique-idx | platform | Results |

Author: Fixed.

|----------------|--------------|----------|---------|
| OFF | OFF |                   | Only the last SRG can contain one last unbounded array for each:<br> t[], u[], b[], s[]|
| OFF | ON  |                   | Only the last SRG can contain one, and only one, last unbounded array across all resource types:<br>Only one of t[], u[], b[], s[]|
| ON  | OFF | DX12              | Each SRG can contain one last unbounded array for each: <br> t[], u[], b[], s[]|

> | ON  | ON  | Vulkan<br> Metal  | Each SRG can contain one, and only one, last unbounded array across all resource types: <br> Only one of t[], u[], b[], s[]|

Author: Fixed.

> ## Things to know before going over some examples:

Reviewer: ```suggestion
## Things to know

To understand the examples on this page, review the following information:
```
* Shorten the title header.
* Move your explanation of this section into the first sentence.

Author: Fixed.

> * "Register Space" in HLSL is the same as "Descriptor Set" in Vulkan.
> * When "--use-spaces" is used, **AZSLc** will assign a unique Register Space per SRG.
> * When "--unique-idx" is used, **AZSLc** will emit all resource descriptors in a Register Space to be in numerical sequence, regardless of the descriptor type. Platforms like Vulkan and Metal require this.
> * Unbounded arrays take over the whole range of register indices from the point they are being declared. This means two things:
>     1. It is possible to declare any kind of resources or any amount of bounded arrays of resources, BEFORE any given Unbounded Array of a particular kind.
>     2. Any resource declared AFTER an Unbounded Array will be considered an error.

Reviewer: Notes for each bullet's edit suggestion:
* Bullet about "register space"
  * Lower case register space and descriptor set. 
  * Italicize "register space" because it is a term that you're defining on this page. 
  * Important: Explain in a sentence or two what register space is? With this, the user can begin to understand from your docs how it is similar to Vulkan. Then, if they wish to learn more, they can conduct further research into Vulkan docs. 
  
    For example, see my suggestion.
* Bullet about `--use-spaces` option    
  * Add context for when we use the `--use-spaces` option. (The context is when we are compiling.)
  * Use in-line code style for command line arguments and options. 
  * Remove bold from AZSLc. Only bold first instance per page. 
  * Lower case register space, as it's written in [Microsoft docs](https://docs.microsoft.com/en-us/windows/win32/direct3d12/resource-binding-in-hlsl)

* Bullet about `--unique-idx` option
  * Same as previous comments.
  * Change "Platforms like Vulkan and Metal..." to "Vulkan and Metal platforms..." to be clear. The phrase, "Platforms like Vulkan and Metal..." indicates that Vulkan and Metal are only some platforms that require this. It makes me  wonder if DirectX requires this too or not? 

* Bullet about unbounded arrays. 
   * Edits to make it read better. 

```suggestion
* In HLSL, a _register space_ is a slot that connects shader data from the CPU to the GPU. For comparison, a register space is similar to a descriptor set in Vulkan. 

* When compiling shader code with the `--use-spaces` option, AZSLc assigns a unique register space per SRG.

* When compiling shader code with the `--unique-idx` option, AZSLc emits all resource descriptors into numerical sequences, regardless of the descriptor type. Vulkan and Metal platforms require this.

* Unbounded arrays take over the whole range of register indices from the point they are declared. This means you must declare any resources including bounded arrays before declaring an unbounded array. Any resource declared after an unbounded array will cause an error.
```

Author: Fixed.

> ## Examples

Reviewer: ### Note about my suggestions that follow
My suggestions in the following examples are to
* Restructure the hierarchy of headings, so that error case examples and solution examples are within the section of the example it's based on.
* Give each example a descriptive name. 

The overall heading structure that I propose looks like this: 

* Examples
  * Example: Register assignments
    * Error case
    * Error case
    * Solution
  * Example: Declaring unbounded arrays
  * Example: Resource assignments with `--unique-idx`
    * Error case
    * Solution
  * Example: Declaring unbounded arrays with `--unique-idx`

Author: Fixed.

> All the examples shown here assume '--use-spaces' is ON when compiling the shader. Which means each SRG will be assigned to its own Register Space.

Reviewer: ```suggestion
All the examples shown here assume that AZSLc compiles the shader code with the `--use-spaces` option set to ON. So, assume that each SRG is assigned to its own register space.
```

Author: Fixed.

> ### Example 1
> In this example there's only one SRG, and no unbounded arrays yet. It is useful to understand register space & index binding assignment.  

Reviewer: ```suggestion
### Example: Register assignments

This example demonstrates how index bindings are assigned to register spaces. It contains one SRG and no unbounded arrays.
```
* Give a more descriptive "Example" title. 
* Swap first and second sentence, so the first sentence explains what the example is about. Then, reword sentence to active voice.

Author: Fixed.

```cpp

>     ShaderResourceGroupSemantic slot1
>     {
>         FrequencyId = 1;
>     };
>     
>     struct MyStruct
>     {
>         float3 m_a;
>         float3 m_b;
>     };
>     
>     //For DX12, --unique-idx is OFF, so the expected register assignment is as follows
>     ShaderResourceGroup SRG1 : slot1
>     {
>         Texture2D<float4>        m_texSRV1;      // Will be bound to t0 register.
>         Texture2D<float4>        m_texSRV2;      // Will be bound to t1 register.
>         RWTexture2D<float4>      m_texUAV1;      // Will be bound to u0 register.
>     	RWTexture2D<float4>      m_texUAV2;      // Will be bound to u1 register.
>         Sampler                  m_sampler1;     // Will be bound to s0 register.
>         Sampler                  m_sampler2;     // Will be bound to s1 register.
>         ConstantBuffer<MyStruct> m_cb1; // Will be bound to b1 register. (Because The ConstantBuffer reserved for SRG1 will be bound to b0 register).
>         ConstantBuffer<MyStruct> m_cb2; // Will be bound to b2 register.
>     };

Reviewer: ```suggestion
ShaderResourceGroupSemantic slot1
{
    FrequencyId = 1;
};

struct MyStruct
{
    float3 m_a;
    float3 m_b;
};

//For DX12, --unique-idx is OFF, so the expected register assignment is as follows
ShaderResourceGroup SRG1 : slot1
{
    Texture2D<float4>        m_texSRV1;     // Binds to t0 register.
    Texture2D<float4>        m_texSRV2;     // Binds to t1 register.
    RWTexture2D<float4>      m_texUAV1;     // Binds to u0 register.
    RWTexture2D<float4>      m_texUAV2;     // Binds to u1 register.
    Sampler                  m_sampler1;    // Binds to s0 register.
    Sampler                  m_sampler2;    // Binds to s1 register.
    ConstantBuffer<MyStruct> m_cb1;         // Binds to b1 register. (The ConstantBuffer reserved for SRG1 binds to b0 register.)
    ConstantBuffer<MyStruct> m_cb2;         // Binds to b2 register.
};
```

Author: Fixed.

```

> ### Example 2: (Error case)

Reviewer: ```suggestion
#### Error case

```
* Remove colon. Give a more descriptive "Example" title.

Author: Fixed.

> This example is based on Example 1, but this time We are trying to introduce the first Unbounded Array: `Texture2D<float4> m_texSRV_unbounded[]`.  

Reviewer: ```suggestion
This example demonstrates how compiling an unbounded array (`Texture2D<float4> m_texSRV_unbounded[]`) results in a register assignment conflict.

```

Author: Fixed.

> This example will fail to compile because, as mentioned before, an Unbounded Array takes over the whole register starting from their index of declaration.  
> In this example, `m_texSRV_unbounded`, is the first texture SRV so it will start at **t0**, and take over the rest of the indices until **tN**. Any other texture SRV declared afterwards won't find a register index to bind to.  

Reviewer: ```suggestion
This example fails to compile because `m_texSRV_unbounded[]` takes over the whole range of register indices from the point its declared. The unbounded array, `m_texSRV_unbounded[]`, is the first texture shader resource view (SRV), so it takes over the whole range of register indices that's allocated for SRVs, from `t0` to `tN`. Texture SRVs that are declared afterwards don't have a register index to bind to.  
```
* Use in-line code style for code elements like `t[ ]` or `tN`. 
* Write out "Shader Resource View (SRV)". For all acronyms across all pages, first introduce the full term and its acronym, then use just the acronym. This is very helpful to readers.

Author: Fixed.

```cpp

>     ShaderResourceGroupSemantic slot1
>     {
>         FrequencyId = 1;
>     };
>     
>     struct MyStruct
>     {
>         float3 m_a;
>         float3 m_b;
>     };
>     
>     //For DX12, --unique-idx is OFF, so the expected register assignment is as follows
>     ShaderResourceGroup SRG1 : slot1
>     {
>         Texture2D<float4>        m_texSRV_unbounded[]; // Will be bound to t0 register, and will take over t1+.
>         Texture2D<float4>        m_texSRV1;      // ERROR. There's no tN available to bind this resource to.
>         Texture2D<float4>        m_texSRV2;      // ERROR. There's no tN available to bind this resource to.
>         RWTexture2D<float4>      m_texUAV1;      // Will be bound to u0 register.
>     	RWTexture2D<float4>      m_texUAV2;      // Will be bound to u1 register.
>         Sampler                  m_sampler1;     // Will be bound to s0 register.
>         Sampler                  m_sampler2;     // Will be bound to s1 register.
>         ConstantBuffer<MyStruct> m_cb1; // Will be bound to b1 register. (Because The ConstantBuffer reserved for SRG1 will be bound to b0 register).
>         ConstantBuffer<MyStruct> m_cb2; // Will be bound to b2 register.
>     };

Reviewer: ```suggestion
ShaderResourceGroupSemantic slot1
{
    FrequencyId = 1;
};

struct MyStruct
{
    float3 m_a;
    float3 m_b;
};

//For DX12, --unique-idx is OFF, so the expected register assignment is as follows
ShaderResourceGroup SRG1 : slot1
{
    Texture2D<float4>        m_texSRV_unbounded[];  // Binds to t0 register, and takes over t1+.
    Texture2D<float4>        m_texSRV1;             // ERROR. There's no tN available to bind this resource to.
    Texture2D<float4>        m_texSRV2;             // ERROR. There's no tN available to bind this resource to.
    RWTexture2D<float4>      m_texUAV1;             // Binds to u0 register.
    RWTexture2D<float4>      m_texUAV2;             // Binds to u1 register.
    Sampler                  m_sampler1;            // Binds to s0 register.
    Sampler                  m_sampler2;            // Binds to s1 register.
    ConstantBuffer<MyStruct> m_cb1;                 // Binds to b1 register. (The ConstantBuffer reserved for SRG1 binds to b0 register).
    ConstantBuffer<MyStruct> m_cb2;                 // Binds to b2 register.
};
```
* Remove indent. It's not necessary when defining within a code snippet (```cpp). This also allows longer comments to fit on the page.
* Simplified "will be bound to" to "binds to".
* Simplified "will take over" to "takes over"

Author: Fixed.

```

> ### Example 3: (Error case)

Reviewer: ```suggestion
#### Error case
```

Author: Fixed.

> This error case is very similar to Example 2, but `m_texSRV_unbounded` is declared after `m_texSRV1`, but before `m_texSRV2`.  

Reviewer: * Add a sentence to briefly but clearly describe the purpose of this example. Tell readers explicitly what the conflict is. Consider my suggestion: 
```suggestion
This example demonstrates a similar register assignment conflict, but defines the unbounded array, `m_texSRV_unbounded`, after `m_texSRV1` and before `m_texSRV2`.  Since `m_texSRV_unbounded` is defined after `m_texSRV1`, `m_texSRV1` can successfully bind to a register. However, since `m_texSRV_unbounded` is defined before `m_texSRV2`, it again allocates a whole range of register indices, leaving `m_texSRV2` without an available resource to bind to. 
```

Author: Fixed.

```cpp

>     ShaderResourceGroupSemantic slot1
>     {
>         FrequencyId = 1;
>     };
>     
>     struct MyStruct
>     {
>         float3 m_a;
>         float3 m_b;
>     };
>     
>     //For DX12, --unique-idx is OFF, so the expected register assignment is as follows
>     ShaderResourceGroup SRG1 : slot1
>     {
>         Texture2D<float4>        m_texSRV1;      // Will be bound to t0 register.
>         Texture2D<float4>        m_texSRV_unbounded[]; // Will be bound to t1 register, and will take over t2+.
>         Texture2D<float4>        m_texSRV2;      // ERROR. There's no tN available to bind this resource to.
>         RWTexture2D<float4>      m_texUAV1;      // Will be bound to u0 register.
>     	RWTexture2D<float4>      m_texUAV2;      // Will be bound to u1 register.
>         Sampler                  m_sampler1;     // Will be bound to s0 register.
>         Sampler                  m_sampler2;     // Will be bound to s1 register.
>         ConstantBuffer<MyStruct> m_cb1; // Will be bound to b1 register. (Because The ConstantBuffer reserved for SRG1 will be bound to b0 register).
>         ConstantBuffer<MyStruct> m_cb2; // Will be bound to b2 register.
>     };

Reviewer: ```suggestion
ShaderResourceGroupSemantic slot1
{
    FrequencyId = 1;
};

struct MyStruct
{
    float3 m_a;
    float3 m_b;
};

//For DX12, --unique-idx is OFF, so the expected register assignment is as follows
ShaderResourceGroup SRG1 : slot1
{
    Texture2D<float4>        m_texSRV1;             // Binds to t0 register.
    Texture2D<float4>        m_texSRV_unbounded[];  // Binds to t1 register, and takes over t2+.
    Texture2D<float4>        m_texSRV2;             // ERROR. There's no tN available to bind this resource to.
    RWTexture2D<float4>      m_texUAV1;             // Binds to u0 register.
    RWTexture2D<float4>      m_texUAV2;             // Binds to u1 register.
    Sampler                  m_sampler1;            // Binds to s0 register.
    Sampler                  m_sampler2;            // Binds to s1 register.
    ConstantBuffer<MyStruct> m_cb1;                 // Binds to b1 register. (The ConstantBuffer reserved for SRG1 binds to b0 register).
    ConstantBuffer<MyStruct> m_cb2;                 // Binds to b2 register.
};
```
* Remove indent. It's not necessary when defining within a code snippet (```cpp). This also allows longer comments to fit on the page.
* Simplified "will be bound to" to "binds to".

Author: Fixed.

```

> ### Example 4:
> This case fixes the issues in Example 3. By moving the point of declaration of `m_texSRV_unbounded[]` after `m_texSRV2` there are no more register assignment conflicts.  

Reviewer: * Add a sentence to teach readers what the rule is that resolves the problem. In this case, it's moving `m_texSRV_unbounded[]` after `m_texSRV2`, but what about in my own shader code? What is the rule I apply? 
  Consider my suggestion for answering that: 

```suggestion
#### Solution

This example demonstrates how to resolve the register assignment conflicts in the previous error cases. To resolve this, define the unbounded array last among the group of resources with the same resource type. In this case, define `m_texSRV_unbounded` after `m_texSRV1` and `m_texSRV2`.
```

```cpp

>     ShaderResourceGroupSemantic slot1
>     {
>         FrequencyId = 1;
>     };
>     
>     struct MyStruct
>     {
>         float3 m_a;
>         float3 m_b;
>     };
>     
>     //For DX12, --unique-idx is OFF, so the expected register assignment is as follows
>     ShaderResourceGroup SRG1 : slot1
>     {
>         Texture2D<float4>        m_texSRV1;      // Will be bound to t0 register.
>         Texture2D<float4>        m_texSRV2;      // Will be bound to t1 register.
>         Texture2D<float4>        m_texSRV_unbounded[]; // Will be bound to t2 register, and will take over t3+.
>         RWTexture2D<float4>      m_texUAV1;      // Will be bound to u0 register.
>     	RWTexture2D<float4>      m_texUAV2;      // Will be bound to u1 register.
>         Sampler                  m_sampler1;     // Will be bound to s0 register.
>         Sampler                  m_sampler2;     // Will be bound to s1 register.
>         ConstantBuffer<MyStruct> m_cb1; // Will be bound to b1 register. (Because The ConstantBuffer reserved for SRG1 will be bound to b0 register).
>         ConstantBuffer<MyStruct> m_cb2; // Will be bound to b2 register.
>     };

Reviewer: ```suggestion
ShaderResourceGroupSemantic slot1
{
    FrequencyId = 1;
};

struct MyStruct
{
    float3 m_a;
    float3 m_b;
};

//For DX12, --unique-idx is OFF, so the expected register assignment is as follows
ShaderResourceGroup SRG1 : slot1
{
    Texture2D<float4>        m_texSRV1;             // Binds to t0 register.
    Texture2D<float4>        m_texSRV2;             // Binds to t1 register.
    Texture2D<float4>        m_texSRV_unbounded[];  // Binds to t2 register, and takes over t3+.
    RWTexture2D<float4>      m_texUAV1;             // Binds to u0 register.
    RWTexture2D<float4>      m_texUAV2;             // Binds to u1 register.
    Sampler                  m_sampler1;            // Binds to s0 register.
    Sampler                  m_sampler2;            // Binds to s1 register.
    ConstantBuffer<MyStruct> m_cb1;                 // Binds to b1 register. (Because the ConstantBuffer reserved for SRG1 binds to b0 register).
    ConstantBuffer<MyStruct> m_cb2;                 // Binds to b2 register.
};
```
* Remove indent. It's not necessary when defining within a code snippet (```cpp). This also allows longer comments to fit on the page.
* Simplified "will be bound to" to "binds to".

Author: Fixed.

```

> ### Example 5:
> This case shows how to declare one Unbounded Array for each resource type. The rule of thumb is to always declare each Unbounded Array after the last non-Unbounded resource.  

Reviewer: ```suggestion
### Example: Declaring unbounded arrays

This case shows how to declare an unbounded array for each resource type. Always declare each unbounded array after the last non-unbounded resource of the same resource type.  
```

Author: Fixed.

```cpp

>     ShaderResourceGroupSemantic slot1
>     {
>         FrequencyId = 1;
>     };
>     
>     struct MyStruct
>     {
>         float3 m_a;
>         float3 m_b;
>     };
>     
>     //For DX12, --unique-idx is OFF, so the expected register assignment is as follows
>     ShaderResourceGroup SRG1 : slot1
>     {
>         Texture2D<float4>        m_texSRV1;      // Will be bound to t0 register.
>         Texture2D<float4>        m_texSRV_bounded[3];  // Will be bound to t1 register... Until t3 (inclusive).
>         Texture2D<float4>        m_texSRV_unbounded[]; // Will be bound to t4 register, and will take over t5+.
>         RWTexture2D<float4>      m_texUAV_bounded[5];  // Will be bound to u0 register... Until u4 (inclusive).
>     	RWTexture2D<float4>      m_texUAV_unbounded[]; // Will be bound to u5 register, and will take over u6+.
>         Sampler                  m_sampler1_bounded[7]; // Will be bound to s0 register... Until s6 (inclusive).
>         Sampler                  m_sampler2_bounded[3]; // Will be bound to s7 register... Until s9 (inclusive).
>         Sampler                  m_sampler_unbounded[]; // Will be bound to s10 register, and will take over s11+.
>         ConstantBuffer<MyStruct> m_cb_array[2]; // Will be bound to b1 register. (Because The ConstantBuffer reserved for SRG1 will be bound to b0 register)... Until     b2 (inclusive).
>         ConstantBuffer<MyStruct> m_cb2; // Will be bound to b3 register.
>         ConstantBuffer<MyStruct> m_cb_unbounded[]; // Will be bound to b4 register, and will take over b5+.
>     };

Reviewer: ```suggestion
ShaderResourceGroupSemantic slot1
{
    FrequencyId = 1;
};

struct MyStruct
{
    float3 m_a;
    float3 m_b;
};

//For DX12, --unique-idx is OFF, so the expected register assignment is as follows
ShaderResourceGroup SRG1 : slot1
{
    Texture2D<float4>        m_texSRV1;             // Binds to t0 register.
    Texture2D<float4>        m_texSRV_bounded[3];   // Binds from t1 to t3 (inclusive).
    Texture2D<float4>        m_texSRV_unbounded[];  // Binds to t4 register and takes over t5+.
    RWTexture2D<float4>      m_texUAV_bounded[5];   // Binds from u0 to u4 (inclusive).
    RWTexture2D<float4>      m_texUAV_unbounded[];  // Binds to u5 register and takes over u6+.
    Sampler                  m_sampler1_bounded[7]; // Binds from s0 to s6 (inclusive).
    Sampler                  m_sampler2_bounded[3]; // Binds from s7 to s9 (inclusive).
    Sampler                  m_sampler_unbounded[]; // Binds to s10 register and takes over s11+.
    ConstantBuffer<MyStruct> m_cb_array[2];         // Binds to b1 register.
    ConstantBuffer<MyStruct> m_cb2;                 // Binds to b3 register.
    ConstantBuffer<MyStruct> m_cb_unbounded[];      // Binds to b4 register, and takes over b5+.
};
```

Author: Fixed.

>     ShaderResourceGroupSemantic slot1
>     {
>         FrequencyId = 1;
>     };
>     
>     struct MyStruct
>     {
>         float3 m_a;
>         float3 m_b;
>     };
>     
>     //For DX12, --unique-idx is OFF, so the expected register assignment is as follows
>     ShaderResourceGroup SRG1 : slot1
>     {
>         Texture2D<float4>        m_texSRV1;      // Will be bound to t0 register.
>         Texture2D<float4>        m_texSRV_bounded[3];  // Will be bound to t1 register... Until t3 (inclusive).
>         Texture2D<float4>        m_texSRV_unbounded[]; // Will be bound to t4 register, and will take over t5+.
>         RWTexture2D<float4>      m_texUAV_bounded[5];  // Will be bound to u0 register... Until u4 (inclusive).
>     	RWTexture2D<float4>      m_texUAV_unbounded[]; // Will be bound to u5 register, and will take over u6+.
>         Sampler                  m_sampler1_bounded[7]; // Will be bound to s0 register... Until s6 (inclusive).
>         Sampler                  m_sampler2_bounded[3]; // Will be bound to s7 register... Until s9 (inclusive).
>         Sampler                  m_sampler_unbounded[]; // Will be bound to s10 register, and will take over s11+.
>         ConstantBuffer<MyStruct> m_cb_array[2]; // Will be bound to b1 register. (Because The ConstantBuffer reserved for SRG1 will be bound to b0 register)... Until     b2 (inclusive).
>         ConstantBuffer<MyStruct> m_cb2; // Will be bound to b3 register.
>         ConstantBuffer<MyStruct> m_cb_unbounded[]; // Will be bound to b4 register, and will take over b5+.
>     };

Reviewer: ```suggestion
ShaderResourceGroupSemantic slot1
{
    FrequencyId = 1;
};

struct MyStruct
{
    float3 m_a;
    float3 m_b;
};

//For DX12, --unique-idx is OFF, so the expected register assignment is as follows
ShaderResourceGroup SRG1 : slot1
{
    Texture2D<float4>        m_texSRV1;             // Binds to t0 register.
    Texture2D<float4>        m_texSRV_bounded[3];   // Binds from t1 to t3 (inclusive).
    Texture2D<float4>        m_texSRV_unbounded[];  // Binds to t4 register and takes over t5+.
    RWTexture2D<float4>      m_texUAV_bounded[5];   // Binds from u0 to u4 (inclusive).
    RWTexture2D<float4>      m_texUAV_unbounded[];  // Binds to u5 register and takes over u6+.
    Sampler                  m_sampler1_bounded[7]; // Binds from s0 to s6 (inclusive).
    Sampler                  m_sampler2_bounded[3]; // Binds from s7 to s9 (inclusive).
    Sampler                  m_sampler_unbounded[]; // Binds to s10 register and takes over s11+.
    ConstantBuffer<MyStruct> m_cb_array[2];         // Binds to b1 register. (The ConstantBuffer reserved for SRG1 binds to b0 register).
    ConstantBuffer<MyStruct> m_cb2;                 // Binds to b3 register.
    ConstantBuffer<MyStruct> m_cb_unbounded[];      // Binds to b4 register, and takes over b5+.
};
```
* Moved comment: "(Because the ConstantBuffer reserved for….)" to simplify the comments for each resource declaration. 
* Remove indent. It's not necessary when defining within a code snippet (```cpp). This also allows longer comments to fit on the page.
* Simplified "will be bound to" to "binds to".
* Simplified "will take over" to "takes over"

Author: Fixed.

```
Now, let's review what will happen to all the examples above when using '--unique-idx', which is the case for Vulkan & Metal.

> ### Example 6: Based on Example 1, with '--unique-idx'
> Same code as Example 1, but it is assumed to be compiled with '--unique-idx' (ON), which is required by Vulkan & Metal. Please note how the register index always increments, regardless of the resource type.  

Reviewer: ```suggestion
### Example: Resource assignments with `--unique-idx`

This example is similar to [Example: Register assignments](#example-register-assignments), but assumes that AZSLc compiles with the `--unique-idx` set to `ON`, which is required by Vulkan and Metal. With the `--unique-idx` option enabled, the register index always increments, regardless of the resource type. 
```

Author: Fixed.

```cpp

>     ShaderResourceGroupSemantic slot1
>     {
>         FrequencyId = 1;
>     };
>     
>     struct MyStruct
>     {
>         float3 m_a;
>         float3 m_b;
>     };
>     
>     //For Vulkan & Metal, --unique-idx is ON, so the expected register assignment is as follows
>     ShaderResourceGroup SRG1 : slot1
>     {
>         Texture2D<float4>        m_texSRV1;      // Will be bound to t1 register. (Because The ConstantBuffer reserved for SRG1 will be bound to b0 register).
>         Texture2D<float4>        m_texSRV2;      // Will be bound to t2 register.
>         RWTexture2D<float4>      m_texUAV1;      // Will be bound to u3 register.
>     	RWTexture2D<float4>      m_texUAV2;      // Will be bound to u4 register.
>         Sampler                  m_sampler1;     // Will be bound to s5 register.
>         Sampler                  m_sampler2;     // Will be bound to s6 register.
>         ConstantBuffer<MyStruct> m_cb1; // Will be bound to b7 register.
>         ConstantBuffer<MyStruct> m_cb2; // Will be bound to b8 register.
>     };

Reviewer: ```suggestion  
ShaderResourceGroupSemantic slot1
{
    FrequencyId = 1;
};

struct MyStruct
{
    float3 m_a;
    float3 m_b;
};

//For Vulkan & Metal, --unique-idx is ON, so the expected register assignment is as follows
// The ConstantBuffer reserved for SRG1 binds to b0 register.
ShaderResourceGroup SRG1 : slot1
{
    Texture2D<float4>        m_texSRV1;     // Binds to t1 register.
    Texture2D<float4>        m_texSRV2;     // Binds to t2 register.
    RWTexture2D<float4>      m_texUAV1;     // Binds to u3 register.
    RWTexture2D<float4>      m_texUAV2;     // Binds to u4 register.
    Sampler                  m_sampler1;    // Binds to s5 register.
    Sampler                  m_sampler2;    // Binds to s6 register.
    ConstantBuffer<MyStruct> m_cb1;         // Binds to b7 register.
    ConstantBuffer<MyStruct> m_cb2;         // Binds to b8 register.
};
```
* Moved comment: "(Because the ConstantBuffer reserved for….)" to simplify the comments for each resource declaration. 
* Remove indent. It's not necessary when defining within a code snippet (```cpp). This also allows longer comments to fit on the page.
* Simplified "will be bound to" to "binds to".
* Simplified "will take over" to "takes over"

Author: Fixed.

```

> ### Example 7: Based on Example 2, with '--unique-idx'. Error Case.

Reviewer: ```suggestion
#### Error case

This example demonstrates how compiling an unbounded array (`Texture2D<float4> m_texSRV_unbounded[]`) with the `--use-idx` option results in a register assignment conflict. 

This example fails to compile because the unbounded array, `m_texSRV_unbounded[]`, takes over a whole range of register indices, regardless of resource type. `m_texSRV_unbounded[]` takes over `tN`, `uN`, `sN`, and `bN` registers, as opposed to the first error case in [Example: Register assignments](#error-case) where it only takes over `tN` registers.
```

- Simplify title. Move the example's explanation out of the title and into its own sentence. 
- Further explain this example. What do you want to teach the reader with this example? Consider my suggestion.

Author: Fixed.

```cpp

>     ShaderResourceGroupSemantic slot1
>     {
>         FrequencyId = 1;
>     };
>     
>     struct MyStruct
>     {
>         float3 m_a;
>         float3 m_b;
>     };
>     
>     //For Vulkan & Metal, --unique-idx is ON, so the expected register assignment is as follows
>     ShaderResourceGroup SRG1 : slot1
>     {
>         Texture2D<float4>        m_texSRV_unbounded[]; // Will be bound to t1 register(Because The ConstantBuffer reserved for SRG1 will be bound to b0 register), and     will take over ALL1+. 
>         Texture2D<float4>        m_texSRV1;      // ERROR. There's no tN available to bind this resource to.
>         Texture2D<float4>        m_texSRV2;      // ERROR. There's no tN available to bind this resource to.
>         RWTexture2D<float4>      m_texUAV1;      // ERROR. There's no uN available to bind this resource to.
>     	RWTexture2D<float4>      m_texUAV2;      // ERROR. There's no uN available to bind this resource to.
>         Sampler                  m_sampler1;     // ERROR. There's no sN available to bind this resource to.
>         Sampler                  m_sampler2;     // ERROR. There's no sN available to bind this resource to.
>         ConstantBuffer<MyStruct> m_cb1; // ERROR. There's no bN available to bind this resource to.
>         ConstantBuffer<MyStruct> m_cb2; // ERROR. There's no bN available to bind this resource to.
>     };

Reviewer: ```suggestion
ShaderResourceGroupSemantic slot1
{
    FrequencyId = 1;
};

struct MyStruct
{
    float3 m_a;
    float3 m_b;
};

//For Vulkan & Metal, --unique-idx is ON, so the expected register assignment is as follows
// The ConstantBuffer reserved for SRG1 binds to b0 register.
ShaderResourceGroup SRG1 : slot1
{
    Texture2D<float4>        m_texSRV_unbounded[];  // Binds to t1 register, and  take over all 1+ registers.
    Texture2D<float4>        m_texSRV1;             // ERROR. There's no tN available to bind this resource to.
    Texture2D<float4>        m_texSRV2;             // ERROR. There's no tN available to bind this resource to.
    RWTexture2D<float4>      m_texUAV1;             // ERROR. There's no uN available to bind this resource to.
    RWTexture2D<float4>      m_texUAV2;             // ERROR. There's no uN available to bind this resource to.
    Sampler                  m_sampler1;            // ERROR. There's no sN available to bind this resource to.
    Sampler                  m_sampler2;            // ERROR. There's no sN available to bind this resource to.
    ConstantBuffer<MyStruct> m_cb1;                 // ERROR. There's no bN available to bind this resource to.
    ConstantBuffer<MyStruct> m_cb2;                 // ERROR. There's no bN available to bind this resource to.
};
```
* Moved comment: "(Because the ConstantBuffer reserved for….)" to simplify the comments for each resource declaration. 
* Remove indent. It's not necessary when defining within a code snippet (```cpp). This also allows longer comments to fit on the page.
* Simplified "will be bound to" to "binds to".
* Simplified "will take over" to "takes over"

Author: Fixed.

```

> ### Example 8: Fix for Example 7.
> The solution to Example 7, is to move the declaration of `m_texSRV_unbounded` to the end of `SRG1`, basically after `m_cb2`.  
> Alternatively, you could create a second `SRG2`, and move the declaration of `m_texSRV_unbounded` to `SRG2` (As the last resource in `SRG2`, in case there are more resources in `SRG2`.)  

Reviewer: ```suggestion
#### Solution

This example demonstrates how to resolve the register assignment conflict in the previous error case. To resolve this, define the unbounded array after all resource declarations, regardless of their resource types. In this case, define `m_texSRV_unbounded` after `m_cb2`. Alternatively, you can move the declaration of `m_texSRV_unbounded` into a second SRG declaration, `SRG2`. The next example demonstrates multiple SRG declarations in detail. 
```

Author: Fixed.

```cpp

>     ShaderResourceGroupSemantic slot1
>     {
>         FrequencyId = 1;
>     };
>     
>     struct MyStruct
>     {
>         float3 m_a;
>         float3 m_b;
>     };
>     
>     //For Vulkan & Metal, --unique-idx is ON, so the expected register assignment is as follows
>     ShaderResourceGroup SRG1 : slot1
>     {
>         Texture2D<float4>        m_texSRV1;      // Will be bound to t1 register. (Because The ConstantBuffer reserved for SRG1 will be bound to b0 register).
>         Texture2D<float4>        m_texSRV2;      // Will be bound to t2 register.
>         RWTexture2D<float4>      m_texUAV1;      // Will be bound to u3 register.
>     	RWTexture2D<float4>      m_texUAV2;      // Will be bound to u4 register.
>         Sampler                  m_sampler1;     // Will be bound to s5 register.
>         Sampler                  m_sampler2;     // Will be bound to s6 register.
>         ConstantBuffer<MyStruct> m_cb1; // Will be bound to b7 register.
>         ConstantBuffer<MyStruct> m_cb2; // Will be bound to b8 register.
>         Texture2D<float4>        m_texSRV_unbounded[]; // Will be bound to t9 register, and will take over ALL10+.
>     };

Reviewer: ```suggestion
ShaderResourceGroupSemantic slot1
{
    FrequencyId = 1;
};

struct MyStruct
{
    float3 m_a;
    float3 m_b;
};
    
// For Vulkan & Metal, --unique-idx is ON, so the expected register assignment is as follows
// The ConstantBuffer reserved for SRG1 binds to b0 register.
ShaderResourceGroup SRG1 : slot1
{
    Texture2D<float4>        m_texSRV1;             // Binds to t1 register.
    Texture2D<float4>        m_texSRV2;             // Binds to t2 register.
    RWTexture2D<float4>      m_texUAV1;             // Binds to u3 register.
    RWTexture2D<float4>      m_texUAV2;             // Binds to u4 register.
    Sampler                  m_sampler1;            // Binds to s5 register.
    Sampler                  m_sampler2;            // Binds to s6 register.
    ConstantBuffer<MyStruct> m_cb1;                 // Binds to b7 register.
    ConstantBuffer<MyStruct> m_cb2;                 // Binds to b8 register.
    Texture2D<float4>        m_texSRV_unbounded[];  // Binds to t9 register, and takes over all 10+ registers.
};
```
* Moved comment: "(Because the ConstantBuffer reserved for….)" to simplify the comments for each resource declaration. 
* Remove indent. It's not necessary when defining within a code snippet (```cpp). This also allows longer comments to fit on the page.
* Simplified "will be bound to" to "binds to".
* Simplified "will take over" to "takes over"

Author: Fixed.

```

> ### Example 9: Similar to Example 5, but with '--unique-idx'
> In example 5, it is shown how to declare one Unbounded Array for each major resource type: SRV(t), UAV(u), Sampler(s) and CBV(b) inside a single SRG.  
> Unfortunately, when using '--unique-idx', this is impossible to do.  

Reviewer: * Reference the earlier example. Compare it to this example. 
* Join second sentence on the same line as the first sentence, so they are one paragraph. 
* 2nd sentence, switch to active voice.
```suggestion
### Example: Declaring unbounded arrays with `--unique-idx`
The earlier example, [Example: Declaring unbounded arrays](#example-declaring-unbounded-arrays), demonstrates how to declare an unbounded array for each major resource type: SRV(t), UAV(u), Sampler(s) and CBV(b) inside a single SRG. You cannot do this when compiling with the `--use-idx` option. 
```

Author: Fixed.

> The only workaround is to declare more SRGs (in this case three more SRGs) and making sure the Unbounded Array is always the last variable declared inside each SRG.  

Reviewer: ```suggestion
To work around this issue, define additional SRGs per resource type. Since each SRG has their set of register spaces that are defined in a numerical sequence, you can avoid resource assignment conflict by separating resources into different SRGs. As in previous examples, always declare an unbounded array as the last variable inside each SRG. 
```
* Add explanation: Why should the user should declare more SRGs? Consider my suggestion. Please update it for technical accuracy. 
  > Since each SRG has their set of register spaces that are defined in a numerical sequence, you can avoid resource assignment conflict by separating resources into different SRGs.

Author: Fixed.

```cpp

>     ShaderResourceGroupSemantic slot1 { FrequencyId = 1; };
>     ShaderResourceGroupSemantic slot2 { FrequencyId = 2; };
>     ShaderResourceGroupSemantic slot3 { FrequencyId = 3; };
>     ShaderResourceGroupSemantic slot4 { FrequencyId = 4; };
>     
>     struct MyStruct
>     {
>         float3 m_a;
>         float3 m_b;
>     };
>     
>     //For Vulkan & Metal, --unique-idx is ON. Because --use-spaces is ON, SRG1 will be bound to Register Space 0.
>     ShaderResourceGroup SRG1 : slot1
>     {
>         Texture2D<float4>        m_texSRV1;      // Will be bound to t1 register. (Because The ConstantBuffer reserved for SRG1 will be bound to b0 register).
>         Texture2D<float4>        m_texSRV_bounded[3];  // Will be bound to t2 register... Until t4 (inclusive).
>         Texture2D<float4>        m_texSRV_unbounded[]; // Will be bound to t5 register, and will take over ALL6+ for Register Space 0.
>         float m_value; // Declaring this variable, of Fundamental type, after an Unbounded Array is ok, as it is not a resource.
>     };
>     
>     //For Vulkan & Metal, --unique-idx is ON. Because --use-spaces is ON, SRG2 will be bound to Register Space 1.
>     ShaderResourceGroup SRG2 : slot2
>     {
>         RWTexture2D<float4>      m_texUAV_bounded[5];  // Will be bound to u1 register (Because The ConstantBuffer reserved for SRG2 will be bound to b0 register)...     Until u5 (inclusive).
>     	RWTexture2D<float4>      m_texUAV_unbounded[]; // Will be bound to u6 register, and will take over ALL7+ for Register Space 1.
>         bool m_boolValue; // Declaring this variable, of Fundamental type, after an Unbounded Array is ok, as it is not a resource.
>     };
>     
>     //For Vulkan & Metal, --unique-idx is ON. Because --use-spaces is ON, SRG3 will be bound to Register Space 2.
>     ShaderResourceGroup SRG3 : slot3
>     {
>         Sampler                  m_sampler1_bounded[7]; // Will be bound to s1 register (Because The ConstantBuffer reserved for SRG3 will be bound to b0 register)...     Until s7 (inclusive).
>         Sampler                  m_sampler2_bounded[3]; // Will be bound to s8 register... Until s10 (inclusive).
>         Sampler                  m_sampler_unbounded[]; // Will be bound to s11 register, and will take over ALL12+ for Register Space 2.
>         float3x3                 m_matrix; // Declaring this variable, of Fundamental type, after an Unbounded Array is ok, as it is not a resource.
>     };
>     
>     //For Vulkan & Metal, --unique-idx is ON. Because --use-spaces is ON, SRG4 will be bound to Register Space 3.
>     ShaderResourceGroup SRG4 : slot4
>     {
>         ConstantBuffer<MyStruct> m_cb_array[2]; // Will be bound to b1 register. (Because The ConstantBuffer reserved for SRG4 will be bound to b0 register)... Until     b2 (inclusive).
>         ConstantBuffer<MyStruct> m_cb2; // Will be bound to b3 register.
>         ConstantBuffer<MyStruct> m_cb_unbounded[]; // Will be bound to b4 register, and will take over ALL5+ for Register Space 3.
>         int m_intValue; // Declaring this variable, of Fundamental type, after an Unbounded Array is ok, as it is not a resource.
>     };

Reviewer: ```suggestion
ShaderResourceGroupSemantic slot1 { FrequencyId = 1; };
ShaderResourceGroupSemantic slot2 { FrequencyId = 2; };
ShaderResourceGroupSemantic slot3 { FrequencyId = 3; };
ShaderResourceGroupSemantic slot4 { FrequencyId = 4; };

struct MyStruct
{
    float3 m_a;
    float3 m_b;
};

// For Vulkan & Metal, --unique-idx is ON, so the expected register assignment is as follows
// SRG1 binds to register space 0. The ConstantBuffer reserved for SRG1 binds to t0 register. 
ShaderResourceGroup SRG1 : slot1
{
    Texture2D<float4>        m_texSRV1;             // Binds to t1 register.
    Texture2D<float4>        m_texSRV_bounded[3];   // Binds from t2 to t4 register (inclusive).
    Texture2D<float4>        m_texSRV_unbounded[];  // Binds to t5 register, and takes over ALL6+ for register space 0.
    float m_value;                                  // Okay to declare a variable of fundamental type as it is not a resource.
};

// SRG2 binds to register space 1. The ConstantBuffer reserved for SRG2 binds to u0 register. 
ShaderResourceGroup SRG2 : slot2
{
    RWTexture2D<float4>      m_texUAV_bounded[5];   // Binds from u1 to u5 register (inclusive).
    RWTexture2D<float4>      m_texUAV_unbounded[];  // Binds to u6 register and takes over ALL7+ for register space 1.
    bool m_boolValue;                               // Okay to declare a variable of fundamental type as it is not a resource.
};

// SRG3 binds to register space 2. The ConstantBuffer reserved for SRG3 binds to s0 register. 
ShaderResourceGroup SRG3 : slot3
{
    Sampler                  m_sampler1_bounded[7]; // Binds from s1 to s7 register (inclusive).
    Sampler                  m_sampler2_bounded[3]; // Binds from s8 to s10 register (inclusive).
    Sampler                  m_sampler_unbounded[]; // Binds to s11 register, and takes over ALL12+ for register space 2.
    float3x3                 m_matrix;              // Okay to declare a variable of fundamental type as it is not a resource.
};

// SRG4 binds to register space 3. The ConstantBuffer reserved for SRG4 binds to b0 register. 
ShaderResourceGroup SRG4 : slot4
{
    ConstantBuffer<MyStruct> m_cb_array[2];     // Binds from b1 to b2 register.
    ConstantBuffer<MyStruct> m_cb2;             // Binds to b3 register.
    ConstantBuffer<MyStruct> m_cb_unbounded[];  // Binds to b4 register, and takes over ALL5+ for register space 3.
    int m_intValue;                             // Okay to declare a variable of fundamental type as it is not a resource.
};
```
* Moved comment: "(Because the ConstantBuffer reserved for….)" to simplify the comments for each resource declaration. 
* Remove indent. It's not necessary when defining within a code snippet (```cpp). This also allows longer comments to fit on the page.
* Simplified "will be bound to" to "binds to".
* Simplified "will take over" to "takes over"

* Moved comment: "//For Vulkan & Metal…" before SRG1, since it applies to all SRGs in this code snippet. Removes redundancy. 
* For SRG1: Copy/paste typo in the comment? It says that SRG1 is bound to b0, but each resource seems to be bound to tN. 
   * Same comment for SRG2 (uN register space) and SRG3 (sN register space). 
   * SRG4 is correct (bN register space)

Author: Fixed. BUT the ConstantBuffers for SRG1, SRG2, SRG3 and SRG3 all  bind to b0 register.
